### ISSUE TITLE
Wildcard Method Requests

### ISSUE THREAD

**Issue by nick-woodward** [franciscop/server#43]
### Description

The current method implementations don't support catch-all endpoints.

### Requested Syntax

```js
const server = require('server');
const { status } = server.reply;
const { get } = server.router;

server(
  // This should get invoked for every request
  // An alternative syntax would be to support Regular Expressions.
  get('*', (ctx) => {
    return status(200);
  })
);
```

### Current Workaround

```js
const server = require('server');
const { status } = server.reply;

server(
  (ctx) => {
    if (ctx.method !== 'GET') return;

    // Do super cool api stuff
    return status(200);
  })
);
```

### Relevant Code

The following code doesn't support wildcards / catch-all endpoints.

https://github.com/franciscop/server/blob/dd1ad51bcd1cd80f64ff119e75edb51d3b6acbd9/router/generic.js#L21-L23

**Comment by franciscop** [comment]
Wait, really? I just added two more tests just in case but both of these seem to be working properly:

```js
get(ctx => { ... });
get('*', ctx => { ... });
```

An empty route will default to `*`: https://github.com/franciscop/server/blob/dd1ad51bcd1cd80f64ff119e75edb51d3b6acbd9/router/parse.js#L3

`path-to-regex-wrap` will match anything with a `*`. Please correct me if I'm wrong. I created a small example with your code and it works.

**HOWEVER**, the bug might be coming from other place. If you put two `get()` on the same level, only the first one will be invoked:

```js
server(
  get('*', ctx => { console.log('called!'); }),
  get('*', ctx => { console.log('never called'); })
);
```

Might this be the issue? See [this commit](https://github.com/franciscop/server/commit/afb7e115dc7542aa6cf27be6ef698bda72ab2a2d) where I tried to reproduce the bug but couldn't.

**Comment by nick-woodward** [comment]
Thanks for the quick response!

I'm having the same issue reproducing this that you're having.
It happens within my project, but small examples don't seem tfo have the same issue.

I'll look into it further tonight to see if its user error or not.

**Comment by franciscop** [comment]
Sorry that last bit it is not an issue, it is a design decision. Only a single route should be invoked in the decision tree (but sub-routes are okay). This is because the response is the return value, to avoid some bugs where a reply might be attempted to be sent twice.

If you do want to do something like that, maybe this might help:

```js
server(
  get('*', 
    get('/users', ctx => { console.log('Called for users'); }),
    ctx => { console.log('Called for all get requests!'); }
  )
);
```

Note: I'd restructure the code a bit though to make sure things are clear.

**Comment by nick-woodward** [comment]
Agreed, when I said I was having the same issue I was referring to not being able to reproduce it with the example I provided. 

I'll follow up later with either an example that I know reproduces the issue or let you know I'm completely unable to reproduce it so you can close this issue.

**Comment by nick-woodward** [comment]
Interesting, after digging in a little further I discovered that the regular expression for my project is being generated differently then on `glitch.com`.

**My Project**

```js
const path = require('path-to-regexp');

path.apply(this, ['*', [], {}]); // /^\*(?:\/)?$/i
```

**[glitch.com](https://glitch.com/edit/#!/franciscop-server-43?path=server.js:9:3)**

```js
const path = require('path-to-regexp');

path.apply(this, ['*', [], {}]); // /^(.*)\/?$/i
```

**Comment by nick-woodward** [comment]
Attempting to upgrade my dependencies in `yarn` doesn't fix the issue. 
I even went to the degree of removing my `yarn.lock` and `node_modules` and installing from scratch.

However if I install my dependencies via NPM it works fine.

**Comment by nick-woodward** [comment]
So I think I have a conclusion to this issue.

### `path-to-regexp`

Some pretty prominent breaking changes were made to path matching 
in 1.x and 2.x hence why [`express` is dependent on `0.1.7`](https://github.com/expressjs/express/blob/351396f971280ab79faddcf9782ea50f4e88358d/package.json#L48).

### `path-to-regexp-wrap`

This module utilizes an [asterisk for its `path-to-regexp` dependency](https://github.com/teologov/path-to-regexp-wrap/blob/a4a25941375d36597a007d548371af5814856b98/package.json#L24).

### NPM

Installs `0.1.7` since this is the latest compatible version with `express`.

### Yarn

Installs `v2.1.0` since this is the latest version available.

**Related Issues**

- https://github.com/yarnpkg/yarn/issues/1785

**Comment by franciscop** [comment]
Woah this looks like a really nasty bug of `yarn`/`npm`, specifically `yarn` working differently. I am not even sure how to start fixing this on my end, but I guess migrating manually to https://github.com/pillarjs/path-to-regexp would be the best considering [how small path-to-regexp-wrap is](https://github.com/teologov/path-to-regexp-wrap/blob/master/index.js).

**Comment by franciscop** [comment]
@teologov you might be interested in this issue

**Comment by teologov** [comment]
Hi guys. Yes, I should probably consider to remove the asterisk and migrate the module to version 2.x

**Comment by teologov** [comment]
Please, see https://github.com/teologov/path-to-regexp-wrap/pull/3

**Comment by franciscop** [comment]
I see this as two issues:

1. Right now installation is broken. npm and yarn will install things differently, so this has to be fixed ASAP. It is the number 1 priority right now for server.
2. Update the path-to-regexp version. This, the same as with express, is a breaking change so it won't be likely until 2.0. The 2.0 version is quite far away, with the 1.1 being another big release considering everything.

Thanks @teologov for fixing it (and so quick!), but due to 2 I am not sure when it'll be doable to update it for server :cry:

**Comment by franciscop** [comment]
I upgraded server to 1.0.7, please @nick-woodward let me know if this fixes the issue for you. I fixed the `path-to-regexp` version and hardcoded `path-to-regexp-wrap`. So it should install the right version and work as expected with both yarn and npm.

For version `2.0` of server I don't know if we will depend on express or not; if we do, it'll be very likely `5.0`. I close this as I am really not sure what will happen by then and it's still quite far away.

**Comment by teologov** [comment]
@franciscop I can deliver a specific `1.x` version of `path-to-regexp-wrap` with `path-to-regexp` of version `0.1.7`, so you could rely on it. 

Please, let me know if this works for you.

**Comment by teologov** [comment]
@franciscop I also published `v1.2.0` with `path-to-regexp@0.1.7`. Hopefully, [it](https://github.com/teologov/path-to-regexp-wrap/blob/v-1/package.json#L24) works for you.

**Comment by nick-woodward** [comment]
@franciscop Sorry for not getting back sooner. 

That did indeed fix my issue. Thanks!
